idaes/core/base/property_meta.py

```diff
@@ -97,27 +97,251 @@ def define_metadata(cls, pcm):
         raise NotImplementedError()
 
 
-class UnitNames(object):
-    """Names for recognized units."""
+class UnitSet(object):
+    """
+    Object defining the set of recognised quantities in IDAES and their base units.
+
+    Units of measurement are defined by setting units for the seven base SI quantities
+    (amount, current, length, luminous intensity, mass, temperature and time), from which units
+    for all other quantities are derived. The units of the seven base quantities must be provided
+    when instantiating the UnitSet, otherwise base SI units are assumed.
+
+    Units can be accesses by via either a property on the UnitSet (e.g., UnitSet.TIME) or
+    via an index on the UnitSet (e.g., UnitSet["time"]).
+    """
+
+    base_quantities = [
+        "time",
+        "length",
+        "mass",
+        "amount",
+        "temperature",
+        "current",
+        "luminous_intensity",
+    ]
+
+    def __init__(
+        self,
+        amount: _PyomoUnit = units.mol,
+        current: _PyomoUnit = units.watt,
+        length: _PyomoUnit = units.meter,
+        luminous_intensity: _PyomoUnit = units.candela,
+        mass: _PyomoUnit = units.kilogram,
+        temperature: _PyomoUnit = units.kelvin,
+        time: _PyomoUnit = units.seconds,
+    ):
+        self._time = time
+        self._length = length
+        self._mass = mass
+        self._amount = amount
+        self._temperature = temperature
+        self._current = current
+        self._luminous_intensity = luminous_intensity
+
+        # Check that valid units were assigned
+        for q in self.base_quantities:
+            u = getattr(self, "_" + q)
+            if not isinstance(u, _PyomoUnit):
+                # Check for non-unit inputs from user
+                raise PropertyPackageError(
+                    f"Unrecognized units of measurement for quantity {q} ({u})"
+                )
+
+            # Check for expected dimensionality
+            expected_dimensionality = {
+                "amount": units.mol,
+                "current": units.watt,
+                "length": units.meter,
+                "luminous_intensity": units.candela,
+                "mass": units.kilogram,
+                "temperature": units.kelvin,
+                "time": units.seconds,
```

Comment:
You could simplify this by making the `base_quantities` list this dict.  Reduces redundancy...

Reply:
I kept them separate as this also defines some "units" - I didn't want a user to accidentally take the units from here and use them somewhere else.